### PR TITLE
Branchformer implementation

### PR DESCRIPTION
**Pull request by kli017**
This PR is an implementation of Branchformer for WeNet encoder structure.
Original Paper: [https://arxiv.org/abs/2207.02971](url)
Original Code: [https://github.com/espnet/espnet/tree/master/espnet2/asr/encoder](url)
Detials:

- Branchformer Encoder And Branchformer Encoderlayer

- [x]  enable concat and fixed_ave merge two branchs
- [x]  enable dynamic chunk and JIT export
- [x]  enable learned_ave merge
- [x]  enable stochastic depth

- Experimental validation (on Aishell)

- [x] batch training
- [x] batch inference

### CODE REVIEW

**Comment by kli017** [comment]
@robin1001 @Mddct @pengzhendong Please review

**Comment by robin1001** [comment]
image

Would you please list the difference of these files to which in `transformer/`? Seems there is a lot of code copy, which is not preferred.

**Comment by kli017** [comment]
Reused moudles from transformer.
branchformer
├── cgmlp.py    # cgmlp module in branchformer block
├── encoder.py # branchformer encoder class
└── encoder_layer.py # branchformer encoder layer

**Comment by kli017** [comment]
@robin1001

File: examples/aishell/s0/run.sh [hunk 1]
```diff
@@ -298,4 +298,4 @@ if [ ${stage} -le 9 ] && [ ${stop_stage} -ge 9 ]; then
   # --lfmmi_dir data/local/lfmmi
```
Comment (robin1001) [inline examples/aishell/s0/run.sh:298]: Seems there is no change in this file, please keep it unchanged.

**Comment by robin1001** [comment]
LGTM, great job!

**Review by robin1001 — APPROVED** [review]

**Pull request merged**